wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/JpaWorkspaceActivityDao.java

```diff
@@ -38,71 +40,190 @@
 
   @Override
   public void setExpiration(WorkspaceExpiration expiration) throws ServerException {
-    requireNonNull(expiration, "Required non-null expiration object");
+    setExpirationTime(expiration.getWorkspaceId(), expiration.getExpiration());
+  }
+
+  @Override
+  public void setExpirationTime(String workspaceId, long expirationTime) throws ServerException {
+    requireNonNull(workspaceId, "Required non-null workspace id");
+    withActivity(workspaceId, a -> a.setExpiration(expirationTime));
+  }
+
+  @Override
+  public void removeExpiration(String workspaceId) throws ServerException {
+    requireNonNull(workspaceId, "Required non-null workspace id");
+    withOptionalActivity(workspaceId, a -> a.setExpiration(null));
+  }
+
+  @Override
+  public List<String> findExpired(long timestamp) throws ServerException {
     try {
-      doCreateOrUpdate(expiration);
+      return doFindExpired(timestamp);
     } catch (RuntimeException x) {
       throw new ServerException(x.getLocalizedMessage(), x);
     }
   }
 
   @Override
-  public void removeExpiration(String workspaceId) throws ServerException {
-    requireNonNull(workspaceId, "Required non-null id");
+  @Transactional
```

Comment:
`RuntimeException` is mapped to `ServerException`. In this case, you must override `rollbackOn` for `Transactional` annotation. Feel free to ask detailed explanation if needed.